packages/flutter/lib/src/services/text_editing_value.dart

```diff
@@ -0,0 +1,444 @@
+      baseOffset: position.offset,
+      extentOffset: position.offset,
+      affinity: position.affinity,
+    );
+  }
+
+  /// Return [selection] expanded to the given [TextPosition].
+  ///
+  /// If the given [TextPosition] is inside of [selection], then [selection] is
+  /// returned without change.
+  ///
+  /// The returned selection will always be a strict superset of [selection].
+  ///
+  /// If extentAtIndex is true, then the [TextSelection.extentOffset] will be
+  /// placed at the given index regardless of the original order of it and
+  /// [TextSelection.baseOffset]. Otherwise, their order will be preserved.
+  ///
+  /// See also:
+  ///
+  ///   * [extendSelectionTo], which is similar but only moves
+  ///     [TextSelection.extentOffset].
+  TextSelection expandSelectionTo(TextPosition position, [bool extentAtIndex = false]) {
+    assert(selection != null);
+
+    final int upperOffset = math.min(selection.baseOffset, selection.extentOffset);
+    final int lowerOffset = math.max(selection.baseOffset, selection.extentOffset);
+    if (position.offset >= upperOffset && position.offset <= lowerOffset) {
+      return selection;
+    }
+
+    if (selection.baseOffset <= selection.extentOffset) {
+      if (position.offset <= selection.baseOffset) {
+        if (extentAtIndex) {
+          return selection.copyWith(
+            baseOffset: selection.extentOffset,
+            extentOffset: position.offset,
+            affinity: position.affinity,
+          );
+        }
+        return selection.copyWith(
+          baseOffset: position.offset,
+          affinity: position.affinity,
+        );
+      }
+      return selection.copyWith(
+        extentOffset: position.offset,
+        affinity: position.affinity,
+      );
+    }
+    if (position.offset <= selection.extentOffset) {
+      return selection.copyWith(
+        extentOffset: position.offset,
+        affinity: position.affinity,
+      );
+    }
+    if (extentAtIndex) {
+      return selection.copyWith(
+        baseOffset: selection.extentOffset,
+        extentOffset: position.offset,
+        affinity: position.affinity,
+      );
+    }
+    return selection.copyWith(
+      baseOffset: position.offset,
+      affinity: position.affinity,
+    );
+  }
+
+  /// Keeping [selection]'s [TextSelection.baseOffset] fixed, move the
+  /// [TextSelection.extentOffset] to the given [TextPosition].
+  TextSelection extendSelectionTo(TextPosition position) {
+    assert(selection != null);
+
+    // If the selection's extent is at the position already, then nothing
+    // happens.
+    if (selection.extent == position) {
+      return selection;
+    }
+
+    return selection.copyWith(
+      extentOffset: position.offset,
+      affinity: position.affinity,
+    );
+  }
+
+  /// {@template flutter.services.TextEditingValue.selectAll}
+  /// Select the entire text value.
+  /// {@endtemplate}
+  TextSelection selectAll() {
+    return selection.copyWith(
```

Comment:
Why the `copyWith`? Is it for keeping the original `isDirectional`? (I'm not sure what `isDirectional` is for, none of our platform text input plugins set that to true it seems).

Reply:
Yes, I didn't want to mess with that as I also wasn't sure where it matters.

Reply:
I ended up removing selectAll and doing it all in TextEditingActionTarget. 6e74a3d